File: backend/shared/contracts/zod/validatedCss.mjs
// validatedCssContract — topology-engine → generate.
// Validates css_processed.json. Documents the schema as currently produced
// by resolvedToLegacyCss(); no cleanup.
//
// Important shape notes:
//   - Same root structure as cssRaw (cssVersion, domain, facility, levelsOrSegments,
//     elements, metadata) but topology adds canonical_id on every element,
//     fills geometry.method, propagates topology._<flag> annotations, and the
//     metadata block carries different keys (no extractBuild; adds outputMode,
//     placementZIsAbsolute, sourceFusion, tunnelDecomposition, repairLog, etc.).
//   - `domain` default differs from cssRaw: 'BUILDING' here vs 'UNKNOWN' there
//     (see resolvedToLegacyCss line 327). Documented in claimsContract footgun
//     comment; not a silent bug today, but a footgun.
//
// PR 1 captures the schema warts-included. Each TODO names a follow-up.
//
// Audit trail:
//   write site            — topology-engine/index.mjs:1448-1452 (css_processed)
//   adapter               — topology-engine/v2-adapter.mjs:259 (resolvedToLegacyCss)
//   geometry annotations  — v2-adapter.mjs:299-302 (preserves _<flag> fields)
//   element-level fields  — v2-adapter.mjs:278-321
//   metadata fields       — v2-adapter.mjs:339-356

import { z } from 'zod';

// ─── Closed enums ────────────────────────────────────────────────────────

// Phase 13 PR 3 provenance status values (mirrors claimsSchema.mjs + resolve/schemas.mjs).
const PROVENANCE_STATUS = z.enum([
  'direct', 'inherited_consensus', 'inherited_contested',
  'derived_geometric', 'derived_inferred', 'missing', 'legacy',
]);

// VARIANT: 'ARCH' is legitimate (hospital render).
const DOMAIN = z.enum(['UNKNOWN', 'TUNNEL', 'BUILDING', 'CIVIL', 'MIXED', 'ARCH']);

// resolvedToLegacyCss maps geom.intent → method via INTENT_TO_METHOD
// (extrusion/sweep/mesh/brep). Plus 'EXTRUSION' as fallback. Closed enum.
const GEOMETRY_METHOD = z.enum(['EXTRUSION', 'SWEEP', 'MESH', 'BREP']);

// outputMode in topology output — observed values include HYBRID + a few
// metadata-only modes. Tighten to closed enum.
const OUTPUT_MODE = z.enum(['HYBRID', 'METADATA_ONLY', 'GEOMETRY_ONLY']);

// VARIANT: 'authoring_safe' is the PRESENTATION_SAFE_MODE setting from
// topology (Phase 12A) — legitimate per-render mode. 'visualization' and
// 'analysis' are speculative; not observed yet but kept for forward compat.
// TODO(phase13-cleanup): catalog all observed values; remove unused.
const EXPORT_PROFILE = z.enum(['coordination', 'visualization', 'analysis', 'authoring_safe']);

// ─── Sub-schemas ─────────────────────────────────────────────────────────

const Vec3 = z.object({
  x: z.number(),
  y: z.number(),
  z: z.number(),
}).strict();

const Bbox = z.object({
  min: Vec3,
  max: Vec3,
}).strict();

// VARIANT: `crs` legitimately propagates through from cssRaw.facility.
// The validatedCss schema originally missed this field — adapter doesn't
// add it, but it survives from upstream. Always nullable in practice.
const Facility = z.object({
  name: z.string(),
  type: z.string(),
  description: z.string(),
  units: z.literal('M'),
  crs: z.string().nullable().optional(),
  origin: Vec3,
  axes: z.literal('RIGHT_HANDED_Z_UP'),
}).strict();

// DEBT: same as cssRaw — every element should carry a complete material.
// Adapter fills a default at v2-adapter.mjs:308 but only if `elem.material`
// is wholly absent; partial materials pass through unchanged.
const Material = z.object({
  name: z.string(),
  color: z.array(z.number()).length(3).optional(),
  transparency: z.number().min(0).max(1).optional(),
}).strict();

// VARIANT: same as cssRaw — some element types legitimately have only
// an origin point. Optional permanently.
const Placement = z.object({
  origin: Vec3,
  axis: Vec3.optional(),
  refDirection: Vec3.optional(),
}).strict();

// Geometry from resolvedToLegacyCss(): always has method, profile, depth;
// optionally direction/path/pathPoints/vertices/faces/mesh and any
// _<topology-annotation> fields preserved from upstream. passthrough.
//
// SEMANTIC: order matters in `path` and `pathPoints` arrays — they are
// polyline/curve definitions; reordering changes geometry. Schema cannot
// express this; do not rearrange in topology without a coordinated
// generate change.
const Geometry = z.object({
  method: GEOMETRY_METHOD,
  profile: z.unknown().nullable(),
  depth: z.number().nullable(),
}).passthrough();
// TODO(phase13-cleanup): catalog observed _<flag> field set
// (_geoBehavior, _isTunnelShell, _pathAuthored — v2-adapter.mjs:299) and
// promote to typed optional fields.

const Relationship = z.object({
  type: z.string(),
  target: z.string(),
}).passthrough();
// passthrough because resolvedToLegacyCss spreads `...r` (v2-adapter.mjs:306)
// preserving any extra keys present on upstream relationships.

// Element shape from resolvedToLegacyCss. Required: id, element_key,
// canonical_id, type, name, placement, geometry, relationships, properties,
// material, confidence, source, sourceFile, metadata.
//
// PR 3 will REPLACE the following pre-existing fields with the full
// Phase 13 provenance schema:
//   - confidence (element-level default 0.7) — replaced or augmented
//   - source (free string 'LLM'/'VSM'/etc.)  — replaced
//   - sourceFile (single string or null)      — replaced with sourceFiles list
const Element = z.object({
  id: z.string(),
  element_key: z.string(),
  canonical_id: z.string(),
  type: z.string(),
  semanticType: z.string().optional(),
  name: z.string(),
  placement: Placement.nullable(),
  geometry: Geometry.nullable(),
  container: z.string().nullable(),
  relationships: z.array(Relationship),
  properties: z.record(z.unknown()),
  material: Material,
  // Pre-existing element-level confidence default (v2-adapter.mjs:309).
  // Phase 14 will replace with measurable per-source/per-pass confidence.
  confidence: z.number().min(0).max(1),
  // Free string: 'LLM' default + 'VSM', 'DXF', 'VISION', etc. PR 3 replaces.
  source: z.string(),
  sourceFile: z.string().nullable(),
  // metadata carries topology placement metadata (zAligned, parentSegment,
  // etc.) plus optional .evidence sub-object. passthrough.
  metadata: z.record(z.unknown()),
  // Phase 13 PR 5 — required. Every topology output element must carry provenance.
  provenance: z.object({
    sourceFile: z.string().nullable(),
    sourceFileStatus: PROVENANCE_STATUS,
    sourceFiles: z.array(z.string()),
    stage: z.string(),
    modifications: z.array(z.string()),
  }).strict(),
}).strict();

const LevelOrSegment = z.object({
  id: z.string(),
  type: z.string(),
  name: z.string(),
  elevation_m: z.number(),
  height_m: z.number(),
}).strict();

// Topology block from resolvedToLegacyCss: 6 sub-arrays.
// TODO(phase13-cleanup): item shapes within each array are unspecified
// — catalog from observed renders.
const Topology = z.object({
  nodes: z.array(z.unknown()),
  runs: z.array(z.unknown()),
  junctions: z.array(z.unknown()),
  interfaces: z.array(z.unknown()),
  hosts: z.array(z.unknown()),
  openings: z.array(z.unknown()),
}).strict();

// Metadata from resolvedToLegacyCss: large set of keys, several optional.
// passthrough because topology stages add their own keys.
// VARIANT: duplicatePositions and outOfBounds are legitimate quality
// counters added by topology validation passes (safety.mjs:95). Optional
// because not every render writes them.
const Metadata = z.object({
  modelExtent: z.object({
    x: z.number(),
    y: z.number(),
    z: z.number(),
    elementCount: z.number().int().min(0),
    duplicatePositions: z.number().int().min(0).optional(),
    outOfBounds: z.number().int().min(0).optional(),
  }).strict(),
  safetyWarnings: z.array(z.unknown()),
  exportProfile: EXPORT_PROFILE,
  outputMode: OUTPUT_MODE,
  placementZIsAbsolute: z.boolean(),
  sourceFusion: z.unknown().nullable(),
  interiorSuppression: z.unknown().nullable(),
  tunnelDecomposition: z.unknown().nullable(),
  repairLog: z.array(z.unknown()),
  cssValidationIssues: z.number().int().min(0),
  cssValidationDetails: z.unknown().optional(),
  ambiguousWallProfiles: z.unknown().optional(),
  facilityDimensions: z.array(z.unknown()),
  materialAssignments: z.array(z.unknown()),
  adapterSource: z.literal('resolvedToLegacyCss'),
  resolvedSchemaVersion: z.string(),
}).passthrough();

// ─── Top-level envelope (strict) + cross-field check ─────────────────────

const ValidatedCssEnvelopeBase = z.object({
  cssVersion: z.literal('1.0'),
  domain: DOMAIN,
  facility: Facility,
  levelsOrSegments: z.array(LevelOrSegment),
  elements: z.array(Element),
  topology: Topology,
  metadata: Metadata,
}).strict();

// CALIBRATION FINDING — KNOWN PIPELINE BUG (filed 2026-05-02)
// =============================================================
// The cross-field invariant `metadata.modelExtent.elementCount ===
// elements.length` is *correct in spirit* but BROKEN in production.
// v2-adapter.mjs:340 explicitly sets `elementCount: elements.length`,
// so the adapter is honest. Something downstream of the adapter mutates
// either elements or elementCount but not both:
//   - hospital render (174fdb79...): elementCount=514 vs elements=477 (+37)
//   - tunnel render   (2b8e02f0...): elementCount=297 vs elements=304 (-7)
// Different signs imply the drift comes from multiple post-adapter
// mutations (some passes add elements without bumping the count; others
// remove without decrementing).
//
// CONSUMERS (who reads metadata.modelExtent.elementCount and gets wrong
// data): grep across backend + ui surfaces ZERO read sites currently.
// Field is purely diagnostic — written by adapter, not consumed by the
// pipeline. Pipeline correctness is unaffected today, BUT once the trace
// UI (PR 2 of Release 13.1) surfaces this field, wrong values will
// display in the Pipeline Trace tab. Fix before PR 2 ships if possible,
// otherwise the trace UI will need to display the count-from-elements
// instead.
//
// The contract does NOT enforce this invariant in PR 1 because doing so
// would block every currently-shipping render. Once topology-engine is
// fixed to keep elementCount in sync (likely a one-line mutation in a
// post-adapter pass that mutates css.elements but forgets to update
// metadata.modelExtent.elementCount), re-enable the .refine() below.
//
// TODO(phase13-followup): file a bug, find the post-adapter mutation
// site that drifts the count, fix it, then re-enable the cross-field
// check. The infrastructure to enforce it is preserved as a comment.
//
// // export const validatedCssContract = ValidatedCssEnvelopeBase.refine(
// //   (env) => env.metadata.modelExtent.elementCount === env.elements.length,
// //   {
// //     message: 'metadata.modelExtent.elementCount must equal elements.length',
// //     path: ['metadata', 'modelExtent', 'elementCount'],
// //   },
// // );

export const validatedCssContract = ValidatedCssEnvelopeBase;

export const validatedCssContractMeta = {
  name: 'validatedCssContract',
  producer: 'topology-engine',
  consumer: 'generate',
  artifact: 'css_processed.json',
  version: '0.1.0',
};

File: backend/shared/contracts/pydantic/builting_contracts/validated_css.py
"""validatedCssContract — topology-engine → generate.

Validates css_processed.json. Documents the schema as currently produced
by resolvedToLegacyCss(); no cleanup. See zod/validatedCss.mjs for the
full operational note.

PR 1 captures the schema warts-included.

Audit trail:
  write site            — topology-engine/index.mjs:1448-1452 (css_processed)
  adapter               — topology-engine/v2-adapter.mjs:259 (resolvedToLegacyCss)
  geometry annotations  — v2-adapter.mjs:299-302 (preserves _<flag> fields)
  element-level fields  — v2-adapter.mjs:278-321
  metadata fields       — v2-adapter.mjs:339-356
"""
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, conlist, model_validator


# ─── Closed enums (Literal types) ───────────────────────────────────────

# Phase 13 PR 3 provenance status values (mirrors claimsSchema.mjs PROVENANCE_STATUS).
ProvenanceStatus = Literal[
    "direct", "inherited_consensus", "inherited_contested",
    "derived_geometric", "derived_inferred", "missing", "legacy",
]

# VARIANT: 'ARCH' is legitimate (hospital render).
Domain = Literal["UNKNOWN", "TUNNEL", "BUILDING", "CIVIL", "MIXED", "ARCH"]

# resolvedToLegacyCss INTENT_TO_METHOD + 'EXTRUSION' fallback.
GeometryMethod = Literal["EXTRUSION", "SWEEP", "MESH", "BREP"]

OutputMode = Literal["HYBRID", "METADATA_ONLY", "GEOMETRY_ONLY"]

# VARIANT: 'authoring_safe' is PRESENTATION_SAFE_MODE (Phase 12A).
# 'visualization' and 'analysis' are speculative; not observed yet.
# TODO(phase13-cleanup): catalog observed values; remove unused.
ExportProfile = Literal["coordination", "visualization", "analysis", "authoring_safe"]


# ─── Sub-schemas ────────────────────────────────────────────────────────


class Vec3(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    x: float
    y: float
    z: float


class Bbox(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    min: Vec3
    max: Vec3


# VARIANT: `crs` legitimately propagates from cssRaw.facility through
# the adapter. Always nullable in practice.
class Facility(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    name: str
    type: str
    description: str
    units: Literal["M"]
    crs: Optional[str] = None
    origin: Vec3
    axes: Literal["RIGHT_HANDED_Z_UP"]


# DEBT: same as cssRaw — partial materials pass through the adapter
# unchanged. Adapter only fills a default when material is wholly absent.
class Material(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    name: str
    color: Optional[conlist(float, min_length=3, max_length=3)] = None
    transparency: Optional[float] = Field(default=None, ge=0.0, le=1.0)


# VARIANT: same as cssRaw — origin-only placements are legitimate.
class Placement(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    origin: Vec3
    axis: Optional[Vec3] = None
    refDirection: Optional[Vec3] = None


# Geometry from resolvedToLegacyCss(): always has method, profile, depth;
# optionally direction/path/pathPoints/vertices/faces/mesh and any
# _<topology-annotation> fields preserved from upstream. extra='allow'.
#
# SEMANTIC: order matters in `path` and `pathPoints` arrays — they are
# polyline/curve definitions; reordering changes geometry. Schema cannot
# express this; do not rearrange in topology without a coordinated
# generate change.
class Geometry(BaseModel):
    model_config = ConfigDict(extra="allow", strict=True)
    method: GeometryMethod
    profile: Optional[Any] = None
    depth: Optional[float] = None
# TODO(phase13-cleanup): catalog observed _<flag> field set
# (_geoBehavior, _isTunnelShell, _pathAuthored — v2-adapter.mjs:299) and
# promote to typed optional fields.


class Relationship(BaseModel):
    # extra='allow' because resolvedToLegacyCss spreads `...r` preserving
    # any extra keys present on upstream relationships (v2-adapter.mjs:306).
    model_config = ConfigDict(extra="allow", strict=True)
    type: str
    target: str


class ElementProvenance(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    sourceFile: Optional[str]
    sourceFileStatus: ProvenanceStatus
    sourceFiles: List[str]
    stage: str
    modifications: List[str]


# Element shape from resolvedToLegacyCss. Phase 13 PR 5 — provenance required.
class Element(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    id: str
    element_key: str
    canonical_id: str
    type: str
    semanticType: Optional[str] = None
    name: str
    placement: Optional[Placement]
    geometry: Optional[Geometry]
    container: Optional[str]
    relationships: List[Relationship]
    properties: Dict[str, Any]
    material: Material
    # Pre-existing element-level confidence default (v2-adapter.mjs:309).
    # Phase 14 will replace with measurable per-source/per-pass confidence.
    confidence: float = Field(ge=0.0, le=1.0)
    # Free string: 'LLM' default + 'VSM', 'DXF', 'VISION', etc. PR 3 replaces.
    source: str
    sourceFile: Optional[str]
    # metadata carries topology placement metadata + optional .evidence
    # sub-object. Free-form for now.
    metadata: Dict[str, Any]
    # Phase 13 PR 5 — required. Every topology output element must carry provenance.
    provenance: ElementProvenance


class LevelOrSegment(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    id: str
    type: str
    name: str
    elevation_m: float
    height_m: float


# Topology block from resolvedToLegacyCss: 6 sub-arrays.
# TODO(phase13-cleanup): item shapes within each array are unspecified.
class Topology(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    nodes: List[Any]
    runs: List[Any]
    junctions: List[Any]
    interfaces: List[Any]
    hosts: List[Any]
    openings: List[Any]


# VARIANT: duplicatePositions and outOfBounds are legitimate quality
# counters from topology safety.mjs:95. Optional; not every render writes.
class ModelExtent(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    x: float
    y: float
    z: float
    elementCount: int = Field(ge=0)
    duplicatePositions: Optional[int] = Field(default=None, ge=0)
    outOfBounds: Optional[int] = Field(default=None, ge=0)


# Metadata from resolvedToLegacyCss: large set of keys, several optional.
# extra='allow' because topology stages add their own keys.
class Metadata(BaseModel):
    model_config = ConfigDict(extra="allow", strict=True)
    modelExtent: ModelExtent
    safetyWarnings: List[Any]
    exportProfile: ExportProfile
    outputMode: OutputMode
    placementZIsAbsolute: bool
    sourceFusion: Optional[Any]
    interiorSuppression: Optional[Any]
    tunnelDecomposition: Optional[Any]
    repairLog: List[Any]
    cssValidationIssues: int = Field(ge=0)
    cssValidationDetails: Optional[Any] = None
    ambiguousWallProfiles: Optional[Any] = None
    facilityDimensions: List[Any]
    materialAssignments: List[Any]
    adapterSource: Literal["resolvedToLegacyCss"]
    resolvedSchemaVersion: str


# ─── Top-level envelope (strict) ────────────────────────────────────────


class ValidatedCssContract(BaseModel):
    """CALIBRATION FINDING — KNOWN PIPELINE BUG (filed 2026-05-02)

    The cross-field invariant `metadata.modelExtent.elementCount ==
    len(elements)` is correct in spirit but BROKEN in production.
    v2-adapter.mjs:340 explicitly sets `elementCount: elements.length`,
    so the adapter is honest. Something downstream of the adapter
    mutates either elements or elementCount but not both:
      - hospital render (174fdb79...): elementCount=514 vs elements=477 (+37)
      - tunnel render   (2b8e02f0...): elementCount=297 vs elements=304 (-7)
    Different signs imply drift from multiple post-adapter mutations.

    CONSUMERS: grep across backend + ui surfaces ZERO read sites today.
    Field is purely diagnostic. Pipeline correctness unaffected, BUT
    once the trace UI (PR 2) surfaces this field, wrong values will
    display. Fix before PR 2 ships if possible.

    The contract does NOT enforce this invariant in PR 1 because doing
    so would block every currently-shipping render. Once topology-engine
    is fixed, re-enable the @model_validator below.

    TODO(phase13-followup): find post-adapter mutation site that drifts
    the count, fix it, then re-enable the cross-field check.
    """
    model_config = ConfigDict(extra="forbid", strict=True)
    cssVersion: Literal["1.0"]
    domain: Domain
    facility: Facility
    levelsOrSegments: List[LevelOrSegment]
    elements: List[Element]
    topology: Topology
    metadata: Metadata

    # @model_validator(mode="after")
    # def _element_count_matches(self) -> "ValidatedCssContract":
    #     if self.metadata.modelExtent.elementCount != len(self.elements):
    #         raise ValueError(
    #             "metadata.modelExtent.elementCount must equal len(elements)"
    #         )
    #     return self


VALIDATED_CSS_CONTRACT_META = {
    "name": "validatedCssContract",
    "producer": "topology-engine",
    "consumer": "generate",
    "artifact": "css_processed.json",
    "version": "0.1.0",
}
